include/deal.II/matrix_free/tools.h

```diff
@@ -157,6 +157,179 @@ namespace MatrixFreeTools
+   * invalid cells are treated as boundary faces.
+   *
+   * The class helps to implement "element birth and death technique",
+   * in which context valid cells are alive and invalid cells are dead.
+   */
+  template <int dim,
+            typename Number,
+            typename VectorizedArrayType = VectorizedArray<Number>>
+  class ElementBirthAndDeathMatrixFree
+  {
+  public:
+    /**
+     * Strcut that helps to configure ElementBirthAndDeathMatrixFree.
+     */
+    struct AdditionalData
+    {
+      /**
+       * Constructor.
+       */
+      AdditionalData(const unsigned int dof_index = 0)
+        : dof_index(dof_index)
+      {}
+
+      /**
+       * Index of the DoFHandler within MatrixFree to be used.
+       */
+      unsigned int dof_index;
+    };
+
+    /**
+     * Reinitialize class based on a given MatrixFree instance.
+     * Particularly, the index of the valid FE is determined.
+     */
+    void
+    reinit(const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+           const AdditionalData &additional_data = AdditionalData())
+    {
+      this->matrix_free = &matrix_free;
+
+      std::set<unsigned int> valid_fe_indices;
+
+      const auto &fe_collection =
+        matrix_free.get_dof_handler(additional_data.dof_index)
+          .get_fe_collection();
+
+      for (unsigned int i = 0; i < fe_collection.size(); ++i)
+        if (fe_collection[i].n_dofs_per_cell() > 0)
+          valid_fe_indices.insert(i);
+
+      AssertDimension(valid_fe_indices.size(), 1);
+
+      fe_index_valid = *valid_fe_indices.begin();
+    }
+
+    /**
+     * Loop over all valid cells.
+     *
+     * For the meaning of the parameters see MatrixFree::cell_loop().
+     */
+    template <typename VectorTypeOut, typename VectorTypeIn>
+    void
+    cell_loop(const std::function<void(
+                const MatrixFree<dim, Number, VectorizedArrayType> &,
+                VectorTypeOut &,
+                const VectorTypeIn &,
+                const std::pair<unsigned int, unsigned int>)> &cell_operation,
+              VectorTypeOut &                                  dst,
+              const VectorTypeIn &                             src,
+              const bool zero_dst_vector = false)
+    {
+      const auto ebd_cell_operation = [&](const auto &matrix_free,
+                                          auto &      dst,
+                                          const auto &src,
+                                          const auto  range) {
+        const auto category = matrix_free.get_cell_range_category(range);
+
+        if (category != fe_index_valid)
+          return;
+
+        cell_operation(matrix_free, dst, src, range);
+      };
+
+      matrix_free->template cell_loop<VectorTypeOut, VectorTypeIn>(
+        ebd_cell_operation, dst, src, zero_dst_vector);
+    }
+
+    /**
+     * Loop over all valid cells and faces. Faces between valid and
+     * invalid cells are treated as boundary faces.
```

Comment:
I agree that it's what most people want but it is very different than what we do everywhere else. That's fine with me but that's something we need to think about. Since the face is treated as a boundary, which boundary ID is used?

Reply:
> I agree that it's what most people want but it is very different than what we do everywhere else. That's fine with me but that's something we need to think about. 

That is why it is not part of `MatrixFree` actual. This is helper class tailored for such purposes.

> Since the face is treated as a boundary, which boundary ID is used?

I would say: `numbers::internal_face_boundary_id`. See https://github.com/dealii/dealii/pull/14122